jiant/tasks/lib/qamr.py

```diff
@@ -0,0 +1,73 @@
+import pandas as pd
+from dataclasses import dataclass
+
+from jiant.tasks.lib.templates import span_prediction as span_pred_template
+from jiant.utils.python.io import read_json_lines
+from jiant.utils.retokenize import TokenAligner, MosesTokenizer
+
+
+class QAMRTask(span_pred_template.AbstractSpanPredicationTask):
+    def get_train_examples(self):
+        return self._create_examples(self.train_path, set_type="train")
+
+    def get_val_examples(self):
+        return self._create_examples(self.val_path, set_type="val")
+
+    def get_test_examples(self):
+        return self._create_examples(self.test_path, set_type="test")
+
+    def _create_examples(self, qa_file_path, set_type):
+        wiki_df = pd.read_csv(self.path_dict["wiki_dict"], sep="\t", names=["sent_id", "text"])
+        wiki_dict = {row["sent_id"]: row["text"] for _, row in wiki_df.iterrows()}
+
+        data_df = pd.read_csv(
+            qa_file_path,
+            sep="\t",
+            header=None,
+            names=[
+                "sent_id",
+                "target_ids",
+                "worker_id",
+                "qa_index",
+                "qa_word",
+                "question",
+                "answer",
+                "response1",
+                "response2",
+            ],
+        )
+        data_df["sent"] = data_df["sent_id"].apply(wiki_dict.get)
+
+        examples = []
+        moses_tokenizer = MosesTokenizer()
+        for i, row in data_df.iterrows():
+            # Answer indices are a space-limited list of numbers.
+            # We simply take the min/max of the indices
+            answer_idxs = list(map(int, row["answer"].split()))
+            answer_token_start, answer_token_end = min(answer_idxs), max(answer_idxs)
+            passage_ptb_tokens = row["sent"].split()
+            passage_space_tokens = moses_tokenizer.detokenize_ptb(passage_ptb_tokens).split()
+            passage_space_str = " ".join(passage_space_tokens)
+
+            token_aligner = TokenAligner(source=passage_ptb_tokens, target=passage_space_tokens)
+            ptb_token_idx_to_space_char_idx = token_aligner.U.dot(token_aligner.C)
```

Comment:
> 3\. I used a lot of source_token_idx to target_char_idx mapping in these two tasks. May we should add this kind of functionality to token aligner?

Yeah, moving this code (common to both tasks) into `retokenize` sounds good. You could do something like this:

```
    def project_token_idx_to_space_char_idx(self, idxs: Union[int, Sequence[int]]) -> Sequence[int]:
        """Project source token index to char indices in the *space-separated* target sequence str.

        Takes a list of token indices in the source token sequence, and returns the corresponding
        char idx in the space-separated target sequence string.

        Args:
            idxs (Union[int, Sequence[int]]): source token index(s) to get related target indices.

        Examples:
            >>> source_tokens = ['abc', 'def', 'ghi', 'jkl']
            >>> target_tokens = ['abc', 'd', 'ef', 'ghi', 'jkl']
            >>> ta = TokenAligner(source_tokens, target_tokens)
            >>> src_token_idx = [2, 3]
            >>> print(ta.project_token_idx_to_space_char_idx(src_token_idx))
            [9 10 11 13 14 15]

        Returns:
            List[int] representing the target indices associated with the provided source indices.

        """
        if isinstance(idxs, int):
            idxs = [idxs]
        token_idx_to_space_char_idx = self.U.dot(self.C)
        return token_idx_to_space_char_idx[idxs].nonzero()[1]
```